sdk/eventhub/Azure.Messaging.EventHubs/tests/EventHubConsumerClient/EventHubConsumerClientTests.cs

```diff
@@ -1273,6 +1272,802 @@ await foreach (PartitionEvent partitionEvent in consumer.ReadEventsFromPartition
+            {
+                await foreach (PartitionEvent partitionEvent in consumer.ReadEventsAsync(cancellation.Token))
+                {
+                    ++receivedEvents;
+                    break;
+                }
+            };
+
+            Assert.That(async () => await invoke(), Throws.TypeOf(exception.GetType()), "The enumerator should surface the exception.");
+            Assert.That(cancellation.IsCancellationRequested, Is.False, "The iteration should have completed normally.");
+            Assert.That(receiveCalls, Is.EqualTo(expectedReceiveCalls).Within(partitions.Length), "The retry policy should have been applied.");
+            Assert.That(receivedEvents, Is.EqualTo(0), "No events should have been received.");
+        }
+
+        /// <summary>
+        ///   Verifies functionality of the <see cref="EventHubConsumerClient.ReadEventsAsync" />
+        ///   method.
+        /// </summary>
+        ///
+        [Test]
+        [TestCaseSource(nameof(NonFatalRetriableExceptionTestCases))]
+        public async Task ReadEventsAsyncHonorsRetryPolicyForRetriableExceptions(Exception exception)
+        {
+            var receiveCalls = 0;
+
+            Func<int, TimeSpan?, IEnumerable<EventData>> receiveCallback = (_max, _time) =>
+            {
+                ++receiveCalls;
+                throw exception;
+            };
+
+            var mockRetryPolicy = new Mock<EventHubsRetryPolicy>();
+            var options = new EventHubConsumerClientOptions { RetryOptions = new RetryOptions { CustomRetryPolicy = mockRetryPolicy.Object } };
+            var transportConsumer = new ReceiveCallbackTransportConsumerMock(receiveCallback);
+            var mockConnection = new MockConnection(() => transportConsumer);
+            var consumer = new EventHubConsumerClient(EventHubConsumerClient.DefaultConsumerGroupName, mockConnection, options);
+            var partitions = await consumer.GetPartitionIdsAsync();
+            var receivedEvents = 0;
+
+            mockRetryPolicy
+                .Setup(policy => policy.CalculateRetryDelay(It.IsAny<Exception>(), It.IsAny<int>()))
+                .Returns(default(TimeSpan?));
+
+            using var cancellation = new CancellationTokenSource(TimeSpan.FromSeconds(10));
+
+            Func<Task> invoke = async () =>
+            {
+                await foreach (PartitionEvent partitionEvent in consumer.ReadEventsAsync(cancellation.Token))
+                {
+                    ++receivedEvents;
+                    break;
+                }
+            };
+
+            Assert.That(async () => await invoke(), Throws.TypeOf(exception.GetType()), "The enumerator should surface the exception.");
+            Assert.That(cancellation.IsCancellationRequested, Is.False, "The iteration should have completed normally.");
+            Assert.That(receiveCalls, Is.EqualTo(1).Within(partitions.Length), "The retry policy should have been respected.");
```

Comment:
I would make it `Is.AtLeast(1).And.AtMost(partitions.Length)` so we don't consider `0` and `partitions.Length + 1` right.

Reply:
Another good call.   Better approach for the variance that my cludge.